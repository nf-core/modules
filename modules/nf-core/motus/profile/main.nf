process MOTUS_PROFILE {
    tag "$meta.id"
    label 'process_medium'

    conda (params.enable_conda ? "bioconda::motus=3.0.1" : null)
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/motus:3.0.1--pyhdfd78af_0':
        'quay.io/biocontainers/motus:3.0.1--pyhdfd78af_0' }"

    input:
    tuple val(meta), path(reads)
    path db

    output:
    tuple val(meta), path("*.out"), emit: out
    tuple val(meta), path("*.bam"), optional: true, emit: bam
    tuple val(meta), path("*.mgc"), optional: true, emit: mgc
    tuple val(meta), path("*.log")                , emit: log
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def inputs = reads[0].getExtension() == 'bam' ?
                    "-i ${reads}" :
                    reads[0].getExtension() == 'mgc' ? "-m $reads" :
                        meta.single_end ?
                            "-s $reads" : "-f ${reads[0]} -r ${reads[1]}"
    def refdb = db ? "-db ${db}" : ""
    """
    motus profile \\
        $args \\
        $inputs \\
        $refdb \\
        -t $task.cpus \\
        -n $prefix \\
        -o ${prefix}.out \\
        2> ${prefix}.log

    ## mOTUs version number is not available from command line.
    ## mOTUs save the version number in index database folder.
    ## mOTUs will check the database version is same version as exec version.
    if [ "$db" == "" ]; then
        VERSION=\$(echo \$(motus -h 2>&1) | sed 's/^.*Version: //; s/References.*\$//')
    else
        VERSION=\$(grep motus $db/db_mOTU_versions | sed 's/motus\\t//g')
    fi
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        motus: \$VERSION
    END_VERSIONS
    """
}
