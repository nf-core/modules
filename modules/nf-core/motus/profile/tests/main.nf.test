nextflow_process {
    name "Test Process MOTUS_PROFILE"
    script "../main.nf"
    process "MOTUS_PROFILE"
    tag "modules"
    tag "modules_nfcore"
    tag "motus"
    tag "motus/downloaddb"
    tag "motus/profile"


    test("sarscov2 - illumina paired end [fastq]") {

        setup {
            run("MOTUS_DOWNLOADDB") {
                script "modules/nf-core/motus/downloaddb/main.nf"
                process {
                    """
                    input[0] = Channel.of([
                        file('https://raw.githubusercontent.com/motu-tool/mOTUs/master/motus/downloadDB.py')
                    ])
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true ),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true )
                    ]
                ]
                input[1] =  MOTUS_DOWNLOADDB.out.db
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(
                    path(process.out.out[0][1]).readLines()[3..10],
                    path(process.out.log[0][1]).readLines().last().contains('Reads are filtered for'),
                    process.out.versions,
                ).match()
            }
            )
        }
    }

    test("sarscov2 - illumina paired end [bam]") {

        setup {
            run("MOTUS_DOWNLOADDB") {
                script "modules/nf-core/motus/downloaddb/main.nf"
                process {
                    """
                    input[0] = Channel.of([
                        file('https://raw.githubusercontent.com/motu-tool/mOTUs/master/motus/downloadDB.py')
                    ])
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file( params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true )]
                input[1] =  MOTUS_DOWNLOADDB.out.db
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(
                    path(process.out.out[0][1]).readLines()[3..10],
                    path(process.out.log[0][1]).readLines().last().contains('Reads are filtered for'),
                    process.out.versions
                ).match()
            }
            )
        }
    }

    test("sarscov2 - illumina single end [fastq] - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file( params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true )
                ]
                input[1] =  []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot( process.out ).match() }
            )
        }
    }

    test("sarscov2 - illumina paired end [fastq] - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true ),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true )
                    ]
                ]
                input[1] =  []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot( process.out ).match() }
            )
        }
    }

    test("sarscov2 - illumina paired end [bam] - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file( params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true )]
                input[1] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot( process.out ).match() }
            )
        }
    }
}
