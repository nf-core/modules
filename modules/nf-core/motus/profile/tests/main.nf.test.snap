{
    "sarscov2 - illumina paired end [fastq]": {
        "content": [
            [
                "Leptospira alexanderi [ref_mOTU_v31_00001]\t0.0000000000",
                "Leptospira weilii [ref_mOTU_v31_00002]\t0.0000000000",
                "Chryseobacterium sp. [ref_mOTU_v31_00004]\t0.0000000000",
                "Chryseobacterium gallinarum [ref_mOTU_v31_00005]\t0.0000000000",
                "Chryseobacterium indologenes [ref_mOTU_v31_00006]\t0.0000000000",
                "Chryseobacterium artocarpi/ureilyticum [ref_mOTU_v31_00007]\t0.0000000000",
                "Chryseobacterium jejuense [ref_mOTU_v31_00008]\t0.0000000000",
                "Chryseobacterium sp. G972 [ref_mOTU_v31_00009]\t0.0000000000"
            ],
            false,
            [
                "versions.yml:md5,0119e45bb894d3205c32fffc0c895d3c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.1"
        },
        "timestamp": "2024-09-04T12:54:13.029947"
    },
    "sarscov2 - illumina paired end [fastq] - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.out:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,361e07d24e9faf2fc49f5c3f4e718517"
                ],
                "bam": [
                    
                ],
                "log": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "mgc": [
                    
                ],
                "out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.out:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,361e07d24e9faf2fc49f5c3f4e718517"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-04T07:44:34.702236603"
    },
    "sarscov2 - illumina paired end [bam]": {
        "content": [
            [
                "Leptospira alexanderi [ref_mOTU_v31_00001]\t0.0000000000",
                "Leptospira weilii [ref_mOTU_v31_00002]\t0.0000000000",
                "Chryseobacterium sp. [ref_mOTU_v31_00004]\t0.0000000000",
                "Chryseobacterium gallinarum [ref_mOTU_v31_00005]\t0.0000000000",
                "Chryseobacterium indologenes [ref_mOTU_v31_00006]\t0.0000000000",
                "Chryseobacterium artocarpi/ureilyticum [ref_mOTU_v31_00007]\t0.0000000000",
                "Chryseobacterium jejuense [ref_mOTU_v31_00008]\t0.0000000000",
                "Chryseobacterium sp. G972 [ref_mOTU_v31_00009]\t0.0000000000"
            ],
            false,
            [
                "versions.yml:md5,0119e45bb894d3205c32fffc0c895d3c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.1"
        },
        "timestamp": "2024-09-04T12:56:30.49163"
    },
    "sarscov2 - illumina single end [fastq] - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.out:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,361e07d24e9faf2fc49f5c3f4e718517"
                ],
                "bam": [
                    
                ],
                "log": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "mgc": [
                    
                ],
                "out": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.out:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,361e07d24e9faf2fc49f5c3f4e718517"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-04T07:44:23.635734599"
    },
    "sarscov2 - illumina paired end [bam] - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.out:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,361e07d24e9faf2fc49f5c3f4e718517"
                ],
                "bam": [
                    
                ],
                "log": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "mgc": [
                    
                ],
                "out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.out:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,361e07d24e9faf2fc49f5c3f4e718517"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-04T07:44:44.53280646"
    }
}